src/virtio-net.c

```diff
@@ -0,0 +1,292 @@
+
+    int ret = poll(pollfds, 2, timeout);
+
+    return ret > 0 && (pollfds[0].revents & POLLIN) && (pollfds[1].revents & POLLOUT);
+}
+
+void sigusr1_handler(int signum) {
+    return ;
+}
+
+static void *virtio_net_vq_avail_handler_rx(void *arg){
+    struct virtq *vq = (struct virtq *)arg;
+    struct virtio_net_dev *dev = (struct virtio_net_dev *)vq->dev;
+
+    struct sigaction sa;
+    sa.sa_handler = sigusr1_handler;
+    sigemptyset(&sa.sa_mask);
+    sa.sa_flags = 0;
+    sigaction(SIGUSR1, &sa, NULL);
+
+    while (!__atomic_load_n(&thread_stop, __ATOMIC_RELAXED)){
+        vq->guest_event->flags = VRING_PACKED_EVENT_FLAG_ENABLE;
+        if (virtio_net_virtq_available_rx(dev, -1))
+            virtq_handle_avail(vq);
+    }
+    return NULL;
+}
+
+static void *virtio_net_vq_avail_handler_tx(void *arg)
+{
+    struct virtq *vq = (struct virtq *) arg;
+    struct virtio_net_dev *dev = (struct virtio_net_dev *) vq->dev;
+
+    struct sigaction sa;
+    sa.sa_handler = sigusr1_handler;
+    sigemptyset(&sa.sa_mask);
+    sa.sa_flags = 0;
+    sigaction(SIGUSR1, &sa, NULL);
+
+    while (!__atomic_load_n(&thread_stop, __ATOMIC_RELAXED)) {
+        vq->guest_event->flags = VRING_PACKED_EVENT_FLAG_ENABLE;
+        if (virtio_net_virtq_available_tx(dev, -1))
+            virtq_handle_avail(vq);
+    }
+    return NULL;
+}
+
+static void virtio_net_enable_vq_rx(struct virtq *vq)
+{
+    struct virtio_net_dev *dev = (struct virtio_net_dev *) vq->dev;
+    vm_t *v = container_of(dev, vm_t, virtio_net_dev);
+
+    if (vq->info.enable)
+        return;
+    vq->info.enable = true;
+    vq->desc_ring = 
+        (struct vring_packed_desc *) vm_guest_to_host(v, vq->info.desc_addr);
+    vq->device_event = (struct vring_packed_desc_event *) vm_guest_to_host(
+            v, vq->info.device_addr);
+    vq->guest_event = (struct vring_packed_desc_event *) vm_guest_to_host(
+            v, vq->info.driver_addr);
+    uint64_t addr = virtio_pci_get_notify_addr(&dev->virtio_pci_dev, vq);
+    vm_ioeventfd_register(v, dev->rx_ioeventfd, addr, 
+                          2, 0);
+    pthread_create(&dev->rx_thread, NULL, virtio_net_vq_avail_handler_rx,
+                   (void *) vq);
+}
+
+static void virtio_net_enable_vq_tx(struct virtq *vq)
+{
+    struct virtio_net_dev *dev = (struct virtio_net_dev *) vq->dev;
+    vm_t *v = container_of(dev, vm_t, virtio_net_dev);
+
+    if (vq->info.enable)
+        return;
+    vq->info.enable = true;
+    vq->desc_ring = 
+        (struct vring_packed_desc *) vm_guest_to_host(v, vq->info.desc_addr);
+    vq->device_event = (struct vring_packed_desc_event *) vm_guest_to_host(
+            v, vq->info.device_addr);
+    vq->guest_event = (struct vring_packed_desc_event *) vm_guest_to_host(
+            v, vq->info.driver_addr);
+ 
+    uint64_t addr = virtio_pci_get_notify_addr(&dev->virtio_pci_dev, vq);
+    vm_ioeventfd_register(v, dev->tx_ioeventfd, addr, 
+                          2, 0);
```

Comment:
Why we use a 2 here instead of  `sizeof(struct virtio_pci_notify_data)`/`notify_cap->cap.length`? Also please don't use magic numbers if possible.

Reply:
Because `cap.length` is 4, the lower 2 bytes are used to notify the device from the `rx_virtqueue`, while the upper 2 bytes are used to notify the device from the `tx_virtqueue`.

Therefore, when calling `virtio_pci_get_notify_addr`, the address of the `rx_virtqueue` is `cap.offset` (assuming `notify_off_multiplier=2` and `vq->info.notify_off=0`), and the address of the `tx_virtqueue` is `notify_off_multiplier * notify_off` (assuming `notify_off_multiplier=2` and `vq->info.notify_off=1`). This allows me to register these two addresses for `rx_ioeventfd` and `tx_ioeventfd`.

Reply:
AFAIK, our `virtio-pci` defines a general notify data which is mapped to the `struct virtio_pci_notify_data`.
```
struct virtio_pci_notify_data {
    uint16_t vqn;
    uint16_t next;
} __attribute__((packed));
```
Does it mean virtio-net use the first 2 bytes for rx event and last 2 bytes for tx event on this structure?

Reply:
If we don't set the `notify_off_multiplier`, when `virtio_net_enable_vq_rx` and `virtio_net_enable_vq_tx` call `uint64_t addr = virtio_pci_get_notify_addr(&dev->virtio_pci_dev, vq)`, it causes the RX and TX to get the same notify address. This results in a problem where the ioeventfd is registered at the same address repeatedly.

To fix this, I changed the `notify_off_multiplier` to 2. Now, `rx_virtq.notify_off` is 0 and `tx_virtq.notify_off` is 1, allowing me to register 2 eventfds at these 2 distinct addresses.

Reply:
According to my understanding, all the virtio devices registered through `virtio_pci_init` have a general MMIO space layout(the `struct virtio_pci_config`) and a general notify data(the `struct virtio_pci_notify_data`). As you can see, we map the first 2 bytes of notify data to `vqn` and the last 2 bytes to `next`. But in your case it looks like you are mapping the first 2 bytes of notify data to `rx_vqn` and the last 2 bytes to `tx_vqn`. Am I correct?

Reply:
Sorry for my late. 
According to my understanding, this might be wrong. When calling `virtio_pci_get_notify_addr`, we obtain the address for the driver to notify the device. However, I think the structure layout is not similar to `virtio_pci_notify_data`. 
```c
uint64_t virtio_pci_get_notify_addr(struct virtio_pci_dev *dev,
                                    struct virtq *vq)
{
    uint64_t base = PCI_HDR_READ(dev->pci_dev.hdr,
                                 PCI_BAR_OFFSET(dev->notify_cap->cap.bar), 32);
    uint64_t offset =
        dev->notify_cap->cap.offset +
        dev->notify_cap->notify_off_multiplier * vq->info.notify_off;
    return base + offset;
}
```
My understanding is that through `notify_off_multiplier` and `notify_off`, we can map two notification areas for RX and TX. We then simply poll these two areas for IO events.

![image](https://github.com/sysprog21/kvm-host/assets/106154786/7a6290d3-295f-4a85-91c3-619e882608e5)

Reply:
The host (I think you mean the virtio device because the host emulates the backend hardware). So the question becomes: how does the device notify the virtqueue to do the transfer?

For example, when virtio-net finds that the tap device has data that needs to be transferred, it needs to get an available buffer from rx_virtq. The device reads the data, writes it to the virtqueue buffer, marks the buffer as used, and then uses a notify callback to send an IRQ to the guest.

Similarly, when the virtio-net device finds that the guest has data to transfer (by polling the notify_cap of the virtio device), the guest writes the data to an available virtqueue buffer from tx_virtq and marks it as available. The device can then get the available virtqueue, read the data from this buffer, and write it to the tap device to complete the transfer from guest to host.

Reply:
Based on your thinking, consider we have another new virtio device that needs 3 vq, so you will define `notify_off_multiplier` to 2, and `vq0.notify_off=0,  vq1.notify_off=1, vq2.notify_off=2` to register 3 eventfds at 3 distinct addresses. According to my understanding, it will cause a problem in this case because we only preserve 4 bytes for notification.

The right way I feel is to increase the entry of `struct virtio_pci_notify_data` under `struct virtio_pci_config`, and assign each vq to a different `virtio_pci_notify_data`.

Reply:
> Hmm, that's not really what I'm asking. Thank you for the explanation of these high-level concepts, but I am just confused about how the code runs to emulate these things.
> 
> I mean, as the spec you post here, it says "The driver is sometime required to send an available buffer notification to the device... this notification involves sending the virtqueue number to the device.". Since we are emulating the device, where can the emulator find the virtq number when the driver sends the notification?

Sorry for my misunderstanding,
Actually in my implement, I don't directily use the vqn of notification data.
I just poll the notification address for io event, but emulator can using `dev->config.notify_data.vqn` to find the vqn.

Reply:
> Based on your thinking, consider we have another new virtio device that needs 3 vqs, so you will define `notify_off_multiplier` to 2, and `vq0.notify_off=0, vq1.notify_off=1, vq2.notify_off=2` to register 3 eventfds at 3 distinct addresses. According to my understanding, it will cause a problem in this case because we only preserve 4 bytes for notification.
> 
> So what I'm thinking here is whether to increase the entries of `struct virtio_pci_notify_data` under `struct virtio_pci_config`, and assign each vq to a different `virtio_pci_notify_data`. If my above thinking is correct.

Yes, it might be a potential problem. However, when I traced notify_data.vqn, I found that the value is always 0, even when using the ping command (sending a packet should use tx_virtq and the guest should write the data to avail_virtq, so I would expect notify.vqn to be 1).

For increasing the entries in the struct `virtio_pci_notify_data`, I think it would require a redesign of the workflow, including changes to the virtio-pci logic. I need to consider how to implement this improvement effectively.

Reply:
I think the KVM host uses `ioeventfd` for the notification mechanism, so we don't need to use the vqn.

When a Guest performs an IO operation that triggers a vm-exit, KVM first checks if an eventfd is registered for that address. If it is, KVM notifies through the eventfd and then allows the Guest to continue execution. This mechanism is useful for IO requests that are primarily for notifications rather than actual data transfer, such as the virtio-pci available buffer notification, thereby reducing the time overhead.

[ref1](https://patchwork.kernel.org/project/kvm/patch/1251028605-31977-23-git-send-email-avi@redhat.com/)
[ref2](https://patchwork.kernel.org/project/kvm/patch/1251028605-31977-23-git-send-email-avi@redhat.com/)

Reply:
Thank you for patiently explaining to me.  I got the point. With ioeventfd, you never use `virtio_pci_notify_data`, so the notification never truly writes the PCI MMIO space.

I'm just thinking if we are going to use ioeventfd for every notification, maybe we should consider removing `virtio_pci_notify_data`, but it is fine to keep this right now.